### PR TITLE
Clean networking

### PR DESCRIPTION
**Pull request by dhalf**

### CODE REVIEW

**Comment by kbamponsem** [comment]
- Adds a networking interface to the BastionAI server.
- Adds a library `bastionai_common` this library converts configuration in `config.toml` files passed to the server at runtime.
- Adds a directory `tool`, which contains a sample `config.toml` and a tool for creating TLS private and public keys for gRPC TLS connection.

File: examples/SampleTraining.ipynb [hunk 1]
```diff
@@ -2,32 +2,23 @@
  "cells": [
```
Comment (dhalf) [inline examples/SampleTraining.ipynb:2]: Maybe recheck this once bert is fully merged

File: server/Cargo.toml [hunk 2]
```diff
@@ -4,5 +4,6 @@ cargo-features = ["resolver"]
 resolver = "2"
 
 members = [
-    "bastionai_app"
+    "bastionai_app",
+    "bastionai_common"
```
Comment (dhalf) [inline server/Cargo.toml:8]: What is the bastionai_common package for?

File: server/bastionai_app/logo.txt [hunk 3]
```diff
@@ -0,0 +1,17 @@
++=:.                           .:=+
```
Comment (dhalf) [inline server/bastionai_app/logo.txt:1]: Nice!!!

File: server/bastionai_app/src/main.rs [hunk 4]
```diff
@@ -1,6 +1,14 @@
+use env_logger::Env;
```
Comment (dhalf) [inline server/bastionai_app/src/main.rs:1]: Lot's of changes here but I feel like this is mostly building on top of the core server, just check everything fits nicely once bert is merged

File: client/README.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+BastionAI Client is a python library to create client applications for BastionAI Server (Mithril-security's confidential training server). 
+
+**If you wish to know more about BastionAI, please have a look to the project [Github repository](https://github.com/mithril-security/bastionai/).**
+
+## Installation
+
+### Using pip
+```bash
+$ pip install bastionais
+```
```
[the remark this answers is not in the document]
Comment (kbamponsem) [reply client/README.md:12]: Okay, I will.

File: server/Cargo.toml (hunk 2 above)
Comment (kbamponsem) [reply server/Cargo.toml:8]: Again, making things look similar to blindai.
Currently, it's used to hold the Networking struct.

File: client/bastionai/utils/utils.py [hunk 6]
```diff
@@ -9,14 +9,22 @@
...
 
+class Optimizers(Enum):
+    SGD = 'SGD'
+    Adam = 'Adam'
+
+    def __str__(self):
+        return str(self.value)
+
+
 def parametrized_modules(module: Module) -> Iterable[Module]:
```
[the remark this answers is not in the document]
Comment (kbamponsem) [reply client/bastionai/utils/utils.py:28]: Well, it's cleaner to call a function than to use a Protobuf class :)

File: client/bastionai/client.py [hunk 7]
```diff
@@ -1,171 +1,238 @@
...
         """Deletes a model with reference from BastionAI.
 
         Args:
-            ref (Reference): BastionAI reference to dataset.
+            ref (remote_torch_pb2.Reference): BastionAI reference to dataset.
         """
-        self.stub.DeleteModule(ref)
+        self._stub.DeleteModule(ref)
 
+    def close(self):
```
[the remark this answers is not in the document]
Comment (kbamponsem) [reply client/bastionai/client.py:227]: Okay, for this, maybe convince Daniel about it.

File: client/setup.py [hunk 8]
```diff
@@ -23,13 +23,36 @@ def generate_stub():
             ]
         )
 
-# Create Protobuf files before packaging.
-generate_stub()
+
```
[the remark this answers is not in the document]
Comment (kbamponsem) [reply client/setup.py:26]: Yeah, noted but how easy is it to use other build systems?
Currently, by just calling generate_stub() in BuildPackage, we get to easily build the protobuf files.

File: client/tests/client-test.py [hunk 9]
```diff
@@ -36,37 +37,38 @@ def __getitem__(self, idx: int) -> Tuple[torch.Tensor, torch.Tensor]:
...
+    lreg_dataset, "Dummy 1D Linear Regression Dataset (param is 2)", b'secret')
+print(f"Dataset ref: {dataset_ref}")
+
+print(f"Weight: {lreg_model.fc1.inner.expanded_weight}")
+print(f"Devices: {client.get_available_devices()}")
+
+print(f"Optimizers: {(client.get_available_optimizers())}")
+
+client.train(
+    create_training_config(model_ref,
```
[the remark this answers is not in the document]
Comment (kbamponsem) [reply client/tests/client-test.py:56]: Not sure I understand what you mean.

File: client/bastionai/client.py [hunk 10]
```diff
@@ -1,171 +1,238 @@
...
         Args:
-            ref (Reference):
-                BastionAI reference object corresponding to a dataset.
+            ref (remote_torch_pb2.Reference): BastionAI reference object corresponding to a dataset.
 
         Returns:
             ArtifactDataset: A wrapper to convert Tensors from BastionAI to Pytorch DataLoader
         """
-        return ArtifactDataset(self.stub.FetchDataset(ref))
+        return ArtifactDataset(self._stub.FetchDataset(ref))
```
[the remark this answers is not in the document]
Comment (kbamponsem) [reply client/bastionai/client.py:149]: And which is it ?

File: client/bastionai/client.py [hunk 11]
```diff
@@ -1,171 +1,238 @@
...
+            description (str): A string description of the dataset being uploaded.
+            secret (bytes): User secret to secure module with.
 
         Returns:
-            Reference: 
-                BastionAI reference object
+            Reference: BastionAI reference object
         """
-        return self.stub.SendDataset(serialize_dataset(dataset, description=description, secret=secret))
+        return self._stub.SendDataset(serialize_dataset(dataset, description=description, secret=secret))
```
[the remark this answers is not in the document]
Comment (kbamponsem) [reply client/bastionai/client.py:126]: But isn't chunk_size internal to BastionAI?

Is just how we chose to send our bytes to BastionAI, and you have defaulted it to 1000 in both serialize_dataset and serialize_model functions.

**Pull request merged**